massa-bootstrap/src/client.rs

```diff
@@ -186,79 +186,77 @@ fn stream_final_state_and_consensus(
-            massa_trace!(
-                "bootstrap.lib.bootstrap_from_server: No error sent at connection",
-                {}
-            );
-        }
-        Err(e) => return Err(e),
-        Ok(BootstrapServerMessage::BootstrapError { error: err }) => {
-            return Err(BootstrapError::ReceivedError(err))
-        }
-        Ok(msg) => return Err(BootstrapError::UnexpectedServerMessage(msg)),
-    };
-
     // handshake
-    let send_time_uncompensated = MassaTime::now()?;
-    // client.handshake() is not cancel-safe but we drop the whole client object if cancelled => it's OK
-    client.handshake(our_version)?;
+    let send_time_uncompensated = MassaTime::now().map_err(|e| (None, e.into()))?;
+
+    let (mut client, ping) = client
+        .handshake(our_version, cfg.read_error_timeout.to_duration())
+        .map_err(|e| (None, e.into()))?;
 
-    // compute ping
-    let ping = MassaTime::now()?.saturating_sub(send_time_uncompensated);
     if ping > cfg.max_ping {
-        return Err(BootstrapError::GeneralError(
-            "bootstrap ping too high".into(),
+        return Err((
+            Some(client),
+            BootstrapError::GeneralError("bootstrap ping too high".into()),
         ));
     }
 
     // First, clock and version.
-    // client.next() is not cancel-safe but we drop the whole client object if cancelled => it's OK
+    // client.next_timeout() is not cancel-safe but we drop the whole client object if cancelled => it's OK
     let server_time = match client.next_timeout(Some(cfg.read_timeout.into())) {
-        Err(e) => return Err(e),
+        Err(e) => return Err((Some(client), e)),
         Ok(BootstrapServerMessage::BootstrapTime {
             server_time,
             version,
         }) => {
             if !our_version.is_compatible(&version) {
-                return Err(BootstrapError::IncompatibleVersionError(format!(
-                    "remote is running incompatible version: {} (local node version: {})",
-                    version, our_version
-                )));
+                return Err((
+                    Some(client),
+                    BootstrapError::IncompatibleVersionError(format!(
+                        "remote is running incompatible version: {} (local node version: {})",
+                        version, our_version
+                    )),
+                ));
             }
             server_time
         }
         Ok(BootstrapServerMessage::BootstrapError { error }) => {
-            return Err(BootstrapError::ReceivedError(error))
+            return Err((Some(client), BootstrapError::ReceivedError(error)))
         }
-        Ok(msg) => return Err(BootstrapError::UnexpectedServerMessage(msg)),
+        Ok(msg) => return Err((Some(client), BootstrapError::UnexpectedServerMessage(msg))),
     };
 
     // get the time of reception
-    let recv_time = MassaTime::now()?;
+    let recv_time = match MassaTime::now() {
```

Comment:
you can use 
```rust 
let recv_time = MassaTime::now()?; 
// or
let recv_time = MassaTime::now().map_err(...)?;
```

Reply:
I tried it that way, but the borrow checker sees that as a move (I think it doesn't recognize the `?` that comes immediately after).

...I don't like it either. I was going to put a comment there, but that would imply needing it everywhere, which as I'm sure you noticed, would be... repetitive :(

Reply:
that's strange but ok it's not a big change